src/Chapter3.hs

```diff
@@ -480,6 +512,44 @@ After defining the city, implement the following functions:
    and at least 10 living __people__ inside in all houses of the city totally.
 -}
 
+data Castle = Castle String | NoCastle
+  deriving Show
+
+data Wall = Wall | NoWall
+  deriving Show
+
+data CultureBuilding = Church | Library
+  deriving Show
+
+data House = House [Person]
+  deriving Show
+
+newtype Person = Person { personName :: String }
+  deriving Show
+
+data City = City { cityCastle :: Castle
+                 , cityWall :: Wall
+                 , cityCultureBuilding :: CultureBuilding
+                 , cityHouses :: [House]
+                 } deriving Show
+
+buildCastle :: City -> String -> City
+buildCastle city castleName = city { cityCastle = Castle castleName }
+
+buildHouse :: City -> [Person] -> City
+buildHouse city@City {..} people =
+  let nPeople = length people
+  in  if nPeople > 0 && nPeople <= 4
+        then city { cityHouses = House people : cityHouses }
+        else city
```

Comment:
~Interesting approach. And this is another pattern in Haskell :slightly_smiling_face: It is usually called "smart constructor". To avoid building houses with more than 4 people you need to restrict usage of the Constructor and allow to create `House`s only with your `mkHouse` function.
You can read a bit more on the pattern in here: https://kowainik.github.io/posts/haskell-mini-patterns#smart-constructor

Generally, in here, you could restrict that on type level, by using something else than `Int` 🙂

Reply:
Ah I like the pattern I'm seeing. At first I didn't see much of a point of why the examples had data constructors with single fields using `Mk...` and  `un...`, but now I see that this pattern combines nicely with smart constructors.

That makes for a nice way to keep the same code shape between both cases. Very nice :smile: